### PR TITLE
Fix for Issue #306 in Extended Stabilizer Simulator

### PR DESCRIPTION
**Pull request by padraic-padraic**
### Summary

The `extended_stabilizer` method would return incorrect values on simple quantim circuits, see Issue #306, including strings that should have 0 output probability.

This PR implements an alternative method for sampling outputs using the Norm Estimation algorithm [described in the corresponding paper](https://quantum-journal.org/papers/q-2019-09-02-181/). This method returns accurate sampling results and passes the Terra test suite, at the expense of an increase in memory and compute time. For example, using Norm Estimation, the tests required ~800s to complete on my laptop, versus ~16s with the metropolis method.

### Details and comments

I am requesting a review of this PR because the difference in performance and capabilites between the two output methods are quite significant, and can impact the project significantly. Thus, I would like to ask about the 'default' parameter choices to implement. In particular, the two main paramters of interest are:

+ `extended_stabilizer_use_metropolis`
+ `extended_stabilizer_disable_measurement_opt`

Based on the value of these paramters, there are three 'modes' the stabilizer simulator can operate in. In the following, `T` refers to the number of shots `n` to the number of qubits, and `m` to the (largest) number of qubits measured.

1. Metropolis, Measurement Optimisation
     + `extended_stabilizer_use_metropolis`: True
     + `extended_stabilizer_disable_measurement_opt`: False
     + Runtime:  n^{2} + (T-1)n
      In this mode, all samples in a measurement are taken at once from a single mixing of the
      monte carlo sampler. This leads to signficantly reduced runtimes but the method is only
      strictly accurate for circuits with non-zero amplitudes on all output strings.
2. Metropolis, No Measurement Optimisation (Current default)
     + `extended_stabilizer_use_metropolis`: True
     + `extended_stabilizer_disable_measurement_opt`: True
     + Runtime: Tn^{2}
     In this mode, each sample is generated by reinitialising and remixing the monte carlo
     sampler. This method gives good results when there are only a small number
     of non-zero amplitudes, as the reinitialisation reduces the risk of
     becoming stuck in zero amplitudes.
3. Norm Estimation
     + `extended_stabilizer_use_metropolis`: True
     + `extended_stabilizer_disable_measurement_opt`: N/A
     + Runtime: Tn^{3}m^{3}
      In this mode, we use an entirely different method to take samples from the output
      distribution. This method avoids the pitfalls of the Monte Carlo method, but at the
      cost of a significant increase in runtime.
      Before taking a sample, we also need to 'cache' the current stabilizer state decomposition
      and thus this method also doubles the memory footprint of the simulator.

I think retaining the current default is likely sensible, but with documentation for users explaining the alternatives and the proper circumstances for choosing each method.

TODO: 
* [x] Change parameters controlling the simulatr
* [x] Add (skippable) tests for norm estimation
* [x] Choose default paramters
* [x] Ensure all flags are documented
* [x] Add a function to pick a good number of norm estimation samples based on the desired accuracy
* [x] Update the memory usage estimates for the case of norm estimation.
* [x] Fix snapshots

### CODE REVIEW

**Comment by chriseclectic** [comment]
@padraic-padraic thanks for putting in the PR.

I think it might make sense to also combine these options into a single configuration parameter. Something like `"extended_stabilizer_sampling_method"` with 3 allowed values something like `"metropolis", "resampled_metropolis", "norm_estimation" (I'm having a hard time thinking of good name to distinguish the two metropolis types). Adding the description of the methods, the assumptions of the outcome distribution, and the runtime estimates that you have here to the documentation would be very helpful.

I'm also thinking that given these conditions we should remove this simulator method from the "automatic" method selection and only have exact simulation methods chosen from for automatic (ie statevector, density_matrix, stabilizer). This seems safest since it requires some more advanced analysis of the type of circuits and expectation of outcomes to choose appropriate config parameters. (On a tangent with 0.5 release we are going to make the simulator classes configurable so it will be much easier to make a simulator with a fixed method).

**Comment by chriseclectic** [comment]
The documentation should be added to the class string of the `QasmSimulator` python class. You can see the current documentation here: 

https://qiskit.org/documentation/stubs/qiskit.providers.aer.QasmSimulator.html#qiskit.providers.aer.QasmSimulator

**Comment by padraic-padraic** [comment]
I have started working on the default parameters and adding the single method switch as you suggested. Just need to test it and I will add them to this PR

**Comment by chriseclectic** [comment]
@padraic-padraic Just wondering if you are still planning on working on this PR or if we should try and get someone else should take over?

**Comment by padraic-padraic** [comment]
Heya,

Apologies for occasionally going dark, I am still working on this but fitting around other stuff sometimes gets in the way. The issue at the moment is just cleaning up how this change affects e.g. measure sampling. Setting a goal to submit this during Hacktoberfest :)

**Comment by padraic-padraic** [comment]
Would there be interest in me adding an extra set of (skippable) "slow" tests for the NormEstimation case, including the test  circuit from #306 ?

**Comment by chriseclectic** [comment]
Hi @padraic-padraic. More tests is generally good! How long do you expect them to take to run? We can decide to skip them later if CI performance becomes an issue.

**Comment by padraic-padraic** [comment]
Running the full set as in `tests/terra/backends/test_qasm_simulator_extended_stabilizer.py` with 100 shots each using norm estimation takes around 800 seconds at the moment (though some parameters are still being tweaked).

**Comment by padraic-padraic** [comment]
Hi all,

I'm happy with this now! I've added documentation, and a single new test based on the cicuit @yaelbh gave in Issue #306. As the simulator is no-longer automatically chosen, I am happy leaving `resampled_metropolis` as the default for the moment as that will give good general performance, and as such added just this one test using norm estimation.

All that's left is for a review, and then to tag the PR and add a reno note!

File: src/simulators/extended_stabilizer/gates.hpp [hunk 1]
```diff
@@ -169,21 +165,22 @@ U1Sample::U1Sample(double lambda)
...
-      {
-        sample_branch_t(coeff_0, Gates::id),
-        sample_branch_t(coeff_1, Gates::s)
-      };
+      gates[0] = Gates::id;
+      gates[1] = Gates::s;
     }
   }
+  phase_0 = coeff_0/std::abs(coeff_0);
+  phase_1 = coeff_1/std::abs(coeff_1);
```
Comment (vvilpas) [inline src/simulators/extended_stabilizer/gates.hpp:178]: It seems we can have `coeff_?` values of zero and in that cases execution will break. Maybe we want this: 
```suggestion
  phase_0 = std::polar(1.0, std::arg(coeff_0));
  phase_1 = std::polar(1.0, std::arg(coeff_1));
```
With `coeff_? = {0.0, 0.0}` will get `phase_? = {1., 0.}` though. Not sure if that is what we want.

File: src/misc/common_macros.hpp [hunk 2]
```diff
@@ -15,8 +15,13 @@
...
 
-#if defined(__GNUC__) && defined(__x86_64__)
- #define GNUC_AVX2
+#if defined(_MSC_VER)
+ #include <intrin.h>
+#elif defined(__GNUC__) && defined(__x86_64__)
+  #ifndef __PPC64__
+     #define GNUC_AVX2
+     #include <cpuid.h>
+  #endif
```
Comment (vvilpas) [inline src/misc/common_macros.hpp:24]: The inclusion header for AVX2 has been refactored in PR #1017 (already merged to master). So when you merge master to your branch, accept master changes regarding all this AVX2 specific things. All those includes have been moved to the file that actual needs them.

**Review by vvilpas — CHANGES_REQUESTED** [review]
Overall looks good! I've left a few comments to check/address.

A thing I would like to discuss is the declaration of functions using the trailing return type. The thing is that in the rest of the source code we are using the regular way of declaring them. We're not sticking to a code standard though, but it's kind of striking have only a section of the code with this kind of declaration. @atilag @chriseclectic thoughts on this?

And last thing, does the test still take **800s** to run?

File: src/simulators/extended_stabilizer/ch_runner.hpp [hunk 3]
```diff
@@ -416,50 +435,152 @@ void Runner::apply_ccz(uint_t control_1, uint_t control_2, uint_t target, uint_t
...
-      adiag_1[l] |= (a[l][i] & (1ULL << i));
-      if (rng.rand() < 0.5)
-      {
-          adiag_2[l] |= (1ULL << i);
-      }
-    }
+    } // end omp parallel
+    // return ParallelNormEstimate(states_, coefficients_, adiag_1, adiag_2, a, num_threads_);
+    double xi = ParallelNormEstimate(states_, coefficients_, adiag_1, adiag_2, a, num_threads_);
+    xi_samples.push_back(xi);
```
[the remark this answers is not in the document]
Comment (padraic-padraic) [reply src/simulators/extended_stabilizer/ch_runner.hpp:478]: Indeed, not intended. Will resolve this!

File: src/misc/common_macros.hpp (hunk 2 above)
Comment (padraic-padraic) [reply src/misc/common_macros.hpp:24]: Okay, great. Have merged that.

File: src/simulators/extended_stabilizer/extended_stabilizer_state.hpp [hunk 4]
```diff
@@ -692,35 +765,52 @@ void State::probabilities_snapshot(const Operations::Op &op, ExperimentResult &r
...
         {
-          target ^= (1ULL << op.qubits[j]);
+          samples = BaseState::qreg_.metropolis_estimation(metropolis_mixing_steps_, probabilities_snapshot_samples_,
+                                                          rng);
+        }
+        else{
+          for (uint_t s=0; s<probabilities_snapshot_samples_; s++)
+          {
+            uint_t sample = BaseState::qreg_.metropolis_estimation(metropolis_mixing_steps_, rng);
+            samples.push_back(sample);
```
[the remark this answers is not in the document]
Comment (padraic-padraic) [reply src/simulators/extended_stabilizer/extended_stabilizer_state.hpp:786]: Yep, `probabilities_snapshot_samples_` should be `uint_t`. Changed that, and added a reservation.

File: src/simulators/extended_stabilizer/gates.hpp (hunk 1 above)
Comment (padraic-padraic) [reply src/simulators/extended_stabilizer/gates.hpp:178]: Yes, that would be the correct behaviour. I didn't consider this case as if one of the coefficients is zero,we'd be applying a pure Pauli or Clifford which shouldn't require any sampling. But I guess it's always possible for such gates to be entered as `U1` gates in the circuit description?

File: src/simulators/extended_stabilizer/chlib/core.hpp [hunk 5]
```diff
@@ -15,15 +15,15 @@
...
+#include <climits>
+#include <cstdint>
 #include <complex>
 #include <iostream>
 #include <vector>
 
-static const int RE_PHASE[8] = {1, 1, 0, -1, -1, -1, 0, 1};
-static const int IM_PHASE[8] = {0, 1, 1, 1, 0, -1, -1, -1};
+static const std::array<int, 8> RE_PHASE = {1, 1, 0, -1, -1, -1, 0, 1};
+static const std::array<int, 8> IM_PHASE = {0, 1, 1, 1, 0, -1, -1, -1};
```
[the remark this answers is not in the document]
Comment (padraic-padraic) [reply src/simulators/extended_stabilizer/chlib/core.hpp:26]: Yep, the double definition I think was due to the definition in `core.hpp` not being in the `CHSimulator` namespace. I've moved that and removed the declaration from `chstabilizer.hpp`.

File: src/simulators/extended_stabilizer/ch_runner.hpp [hunk 6]
```diff
@@ -416,50 +435,152 @@ void Runner::apply_ccz(uint_t control_1, uint_t control_2, uint_t target, uint_t
...
+  }
+  std::sort(xi_samples.begin(), xi_samples.end(), std::greater<double>());
+  if(repetitions % 2 == 0)
+  {
+  	return (xi_samples[repetitions/2] + xi_samples[repetitions/2 - 1]) / 2.;
+  }
+  else
+  {
+  	uint_t index = std::floor(repetitions / 2.);
+  	return xi_samples[index];
```
[the remark this answers is not in the document]
Comment (padraic-padraic) [reply src/simulators/extended_stabilizer/ch_runner.hpp:493]: Yep, I'm happy with this change. Indeed I think we can leave the sort optimization to the compiler.

**Comment by padraic-padraic** [comment]
I am happy with these suggestions! I have addressed and merged the suggested changes.

With regards to the trailing return type, that change was made a while back (this issue has been open a long time..) by linting my code according to some clang-tidy settings I recall finding in the project? Those I guess haven't been applied across the source so I'm happy to switch it back to the regular definition style if desired.

With regards to tests taking 800s: that's if I run the entire test suite with norm estimation. In fact, the figure is more like 1200s including the proper repeated-norm-estimation-and-return-median process. As I mentioned, given this simulator is no longer automatically suggested I am happy to keep metropolis as the default sampling method, and thus I only added one test that runs with norm estimation. The test suite takes about 60s to run with this additional test.

**Comment by vvilpas** [comment]
Great! 
Regarding trailing return types function definitions, please switch back to the regular definition, to be consistent with the rest of the code. I'll look into that clang-tidy file you mention.
Please, add a release note and we are good to merge!

**Review by vvilpas — APPROVED** [review]

**Pull request merged**